### ISSUE TITLE
"Unknown" pseudo-elements in React Developer Tools

### ISSUE THREAD

**Issue by paulsmith** [jlongster/jsx-reader#1]
Under the [React Developer Tools](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) inspector, JSX compiled with `sjs -l jsx-reader` doesn't display properly in the pseudo DOM tree the inspect renders.

Specifically, the names of React component classes are all replaced with "Unknown".

Expected:
![image](https://cloud.githubusercontent.com/assets/1210/3473420/b57a19c4-02d7-11e4-8dc3-5ca5c200f7da.png)

Actual:
![image](https://cloud.githubusercontent.com/assets/1210/3473433/daa663c4-02d7-11e4-85a1-1aed492f57d7.png)

This is probably actually a bug with Developer Tools, not being able to handle to Sweet.js-mangled hygienic names like "List$751", but jsx-reader surfaced it so thought I'd report it here.

**Comment by jlongster** [comment]
Strange. All of the React.DOM.\* calls aren't mangled because those identifiers are global. I don't know how the developer tools discovers the element names, but I'll definitely look into it. Thanks for reporting!

**Comment by jlongster** [comment]
Ooh, it's just your custom elements. That might be a problem with mangling, I suppose. Try invoking sjs with the "-r" option, which tells it to output readable names. It runs a pass on the output and removes any unnecessary mangling.

**Comment by andreypopp** [comment]
Original `jsx` transform adds `displayName` property to component's spec:

```
var Component = React.createClass({
  displayName: 'Component',
  ...
})
```

This is what is used by React Dev Tools and in error messages.

**Comment by paulsmith** [comment]
@jlongster @andreypopp Both the "-r" sjs option and the displayName property fix the problem. Thanks!

**Comment by jlongster** [comment]
Cool! I don't think you need the `-r` option if the displayName fixes it. What did you change with `displayName` to get it to work? If it worked before but not now, I don't understand because JSX doesn't do anything with `React.createClass`.

**Comment by andreypopp** [comment]
@jlongster JSX doesn't do anything with `createClass` but there's additional transform for that https://github.com/facebook/react/blob/master/vendor/fbtransform/transforms/reactDisplayName.js.

**Comment by andreypopp** [comment]
Some time ago I experimented with Sweet.js and made some trivial macros https://github.com/andreypopp/react-macros to define React components. It adds `displayName` automatically. I don't know if this is good idea or not to define custom syntax for the purpose but it was fun to experiment with nevertheless.

**Comment by jlongster** [comment]
Ah ok. So does JSX do that by default, or do you have to pass the flag?

@paulsmith Did you add `displayName` manually?

**Comment by andreypopp** [comment]
@jlongster `require('react-tools').transform` (and so `jsx` executable) does that by default

**Comment by paulsmith** [comment]
@jlongster Yep, sorry for the confusion: I should have said _either_ adding `displayName` manually, or using `-r` option fixed the problem. (I had not set `displayName` prior; I learned about it from @andreypopp.)

**Comment by jlongster** [comment]
Try the latest version on master. I added a macro to add display name into the component. I'm not publishing it yet because it needs to be tested more (and I need to check if `displayName` is already manually specified)

**Comment by paulsmith** [comment]
Version on master successfully set `displayName` on the same code I used screenshotted above, and looks good in Dev Tools. Thanks!

**Comment by jlongster** [comment]
@paulsmith Great! I just refactored this functionality a bit to fix issues from #2, so if you wouldn't mind, pull down the latest master and make sure it still works. The newest macro also only adds `displayName` if it doesn't exist already.

**Comment by paulsmith** [comment]
Confirmed, working as expected.
